### ISSUE TITLE
go/types: incorrect type reported for comma-err C functions (manifests itself in IDEs)

### ISSUE THREAD

**Issue by AndSDev** [golang/go#47777]
Incorrect type for error in C-binding in IDE.
Go build / run works fine, but in IDE (VS Code) type of `err` is `bool` instead of `error`.

### What version of Go are you using (`go version`)?

<pre>
$ go version
go version go1.17 linux/amd64
</pre>
```
$ gopls version
golang.org/x/tools/gopls v0.7.1
    golang.org/x/tools/gopls@v0.7.1 h1:Mh3Z8Xcoq3Zy7ksSlwDV/nzQSbjFf06A+L+F8YHq55U
```

### Does this issue reproduce with the latest release?
Yes. Only the latest release is affected (1.17).

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/home/dev/.cache/go-build"
GOENV="/home/dev/.config/go/env"
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GOMODCACHE="/home/dev/go/pkg/mod"
GONOPROXY=""
GONOSUMDB=""
GOOS="linux"
GOPATH="/home/dev/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/home/dev/.local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/home/dev/.local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.17"
GCCGO="gccgo"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/home/dev/go/src/sandbox/go.mod"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build3354119137=/tmp/go-build -gno-record-gcc-switches"
</pre></details>

### What did you do?
Open the following in VS Code (main.go):
```go
package main

// int fortytwo() { return 42; }
import "C"
import "fmt"

func main() {
	r, err := C.fortytwo()
	if err != nil {
		panic(err)
	}
	fmt.Println(r)
	// Output: 42
}
```

### What did you expect to see?
Type of `err` is `error`.

### What did you see instead?
Type of `err` is `bool`.
IDE shows error: `cannot convert nil (untyped nil value) to bool`.

![cgo_gopls_diff](https://user-images.githubusercontent.com/13585370/129883752-73390fae-f71c-4c5f-9a32-0f72ae90ec37.png)

**Comment by stamblerre** [comment]
@AndSDev: Does clicking the "regenerate cgo definitions" button fix the problem?

**Comment by AndSDev** [comment]
> @AndSDev: Does clicking the "regenerate cgo definitions" button fix the problem?

"regenerate cgo definitions" does not help.

After switching the go version, VS Code asks to update the tools (in the notification) - after the update there is a problem.
Doing "Go: Install/Update tools" manually does not solve the problem.
"Go: Install/Update tools" allows to select packages to update.
If I manually rebuild only the `gopls` (after switching the go version and without rebuilding everything from the notification), there will be problems.
If I manually rebuild everything without `gopls` (after switching the go version and without rebuilding everything from the notification) - no problem.

**Comment by stamblerre** [comment]
Can you please share your `gopls` logs when you see the issue? Information on how to capture them can be found [here](https://github.com/golang/vscode-go/blob/master/docs/troubleshooting.md#collect-gopls-information).

**Comment by zosmac** [comment]
<img width="899" alt="VSCode source and problems view " src="https://user-images.githubusercontent.com/4759874/129976873-0459ab34-83f8-4358-bbe5-3d94bed7733d.png">

**Comment by zosmac** [comment]
[VSCode gopls output.txt](https://github.com/golang/go/files/7010408/VSCode.gopls.output.txt)

**Comment by zosmac** [comment]
I too am seeing this problem (on macOS/Darwin) with gopls reporting a type mismatch for C function calls, expecting a bool return type rather than errno/error.

Go 1.17,
VSCode 1.59.0
gopls 0.7.1

**Comment by zosmac** [comment]
Interestingly, each function that is complaining about the mismatched error type has the C interface routine defined twice, one with two return values, and one with only one return value.

**Comment by zosmac** [comment]
If I select "Go to Definition" for C.proc_pidfdinfo called in the example above, VSCode takes me to this:

//go:cgo_unsafe_args
func _Cfunc_proc_pidfdinfo(p0 _Ctype_int, p1 _Ctype_int, p2 _Ctype_int, p3 unsafe.Pointer, p4 _Ctype_int) **(r1 _Ctype_int)** {
	_cgo_runtime_cgocall(_cgo_577eabb07159_Cfunc_proc_pidfdinfo, uintptr(unsafe.Pointer(&p0)))
	if _Cgo_always_false {
		_Cgo_use(p0)
		_Cgo_use(p1)
		_Cgo_use(p2)
		_Cgo_use(p3)
		_Cgo_use(p4)
	}
	return
}

**Comment by zosmac** [comment]
However, if I search in the file for proc_pidfdinfo, I also find this:

//go:cgo_import_static _cgo_577eabb07159_C2func_proc_pidfdinfo
//go:linkname __cgofn__cgo_577eabb07159_C2func_proc_pidfdinfo _cgo_577eabb07159_C2func_proc_pidfdinfo
var __cgofn__cgo_577eabb07159_C2func_proc_pidfdinfo byte
var _cgo_577eabb07159_C2func_proc_pidfdinfo = unsafe.Pointer(&__cgofn__cgo_577eabb07159_C2func_proc_pidfdinfo)

//go:cgo_unsafe_args
func _C2func_proc_pidfdinfo(p0 _Ctype_int, p1 _Ctype_int, p2 _Ctype_int, p3 unsafe.Pointer, p4 _Ctype_int) **(r1 _Ctype_int, r2 error)** {
	errno := _cgo_runtime_cgocall(_cgo_577eabb07159_C2func_proc_pidfdinfo, uintptr(unsafe.Pointer(&p0)))
	if errno != 0 { r2 = syscall.Errno(errno) }
	if _Cgo_always_false {
		_Cgo_use(p0)
		_Cgo_use(p1)
		_Cgo_use(p2)
		_Cgo_use(p3)
		_Cgo_use(p4)
	}
	return
}

**Comment by AndSDev** [comment]
I reproduce this issue on a new user (all caches are empty, clear install of go 1.17).
gopls trace with verbose flag (for empty user):
[trace_verb.log](https://github.com/golang/go/files/7012339/trace_verb.log)

**Comment by findleyr** [comment]
This is a regression in go/types in Go 1.17, introduced by me.

Bisected to https://golang.org/cl/282193, where at call.go:305 I checked for commaerr operands, but overlooked that the operand mode had already been set to value on line 303.

This needs to be fixed and backported to 1.17.  @mdempsky might know more about the potential impact of this bug beyond gopls.

CC @griesemer

I have spotty internet right now, so might be slow to respond.

**Comment by mdempsky** [comment]
I think `commaerr` should only affect gopls. The `UsesCgo` flag isn't generally available to other go/types users yet.

**Comment by griesemer** [comment]
I can reproduce this and have a fix. Will send out CL a bit later today.

**Comment by griesemer** [comment]
@gopherbot please backport to 1.17. This is a regression.

**Comment by dmitshur** [comment]
The fix that closed this issue landed into master branch, which is where Go 1.18 development is happening, so moving this issue to the Go1.18 milestone. (The 1.17 backport issue #47854 is in 1.17.1 milestone.)
